From df61514c6e07ae722c87b2727999bdc49a08bcaa Mon Sep 17 00:00:00 2001
From: "Roberto A. Foglietta" <roberto.foglietta@gmail.com>
Date: Tue, 12 May 2026 05:39:41 +0200
Subject: [PATCH] unseeded cnrg warning flooding dmesg fix, v2
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Better than hiding the warning, there is a third approach compared
print_ratelimited() and printk_once(), printk one over N=256 which
is fitting this specific case: at boot time too much warnings for
a few seconds but still printing something in case the problem
persists without clogging the log.

On a SMP system the counter might face a race condition due to non
atomic increment and this can lead to skipping a scheduled print or
print two of them when just one was expected. No harm in both cases,
anyway.

To be honest, complaining that a suppression counter might be subject
to a race condition on an SMP system - when, if it weren’t there, an
alarm would be triggered stating that the system is initialising
cryptographic directives without adequate entropy support - seems
a bit surreal to me.

Patch based on Linux kernel 5.15.202 LTS master branch.

v1 -> v2: a bit better English and "git am" format

Signed-off-by: Roberto A. Foglietta <roberto.foglietta@gmail.com>
---
 drivers/char/random.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/char/random.c b/drivers/char/random.c
index 8642326de6e1..eeec336b10fa 100644
--- a/drivers/char/random.c
+++ b/drivers/char/random.c
@@ -182,7 +182,8 @@ static void __cold process_random_ready_list(void)
 }
 
 #define warn_unseeded_randomness() \
-	if (IS_ENABLED(CONFIG_WARN_ALL_UNSEEDED_RANDOM) && !crng_ready()) \
+	if (IS_ENABLED(CONFIG_WARN_ALL_UNSEEDED_RANDOM) && !crng_ready() \
+	&& !({ static u8 x=-1; ++x; })) \
 		printk_deferred(KERN_NOTICE "random: %s called from %pS with crng_init=%d\n", \
 				__func__, (void *)_RET_IP_, crng_init)
 
-- 
2.34.1

